azimuth_capi/operator.py

```diff
@@ -220,77 +226,88 @@ async def fetch_model_instance(model, name, namespace = None):
     api.Cluster,
     kopf.on.validate,
     # We want to validate the instance ourselves
-    include_instance = False,
-    id = "validate-cluster"
+    include_instance=False,
+    id="validate-cluster",
 )
 async def validate_cluster(name, namespace, meta, spec, operation, **kwargs):
     """
     Validates cluster objects.
     """
     if operation not in {"CREATE", "UPDATE"}:
         return
-    try:
-        spec = api.ClusterSpec.model_validate(spec)
-    except pydantic.ValidationError as exc:
-        raise kopf.AdmissionError(str(exc), code = 400)
-    # The credentials secret must exist, unless the cluster is deleting
+
+    spec = _validate_spec_model(spec)
     if not meta.get("deletionTimestamp"):
-        secrets = await ekclient.api("v1").resource("secrets")
-        try:
-            _ = await secrets.fetch(
-                spec.cloud_credentials_secret_name,
-                namespace = namespace
-            )
-        except ApiError as exc:
-            if exc.status_code == 404:
-                raise kopf.AdmissionError(
-                    "specified cloud credentials secret does not exist",
-                    code = 400
-                )
-            else:
-                raise
-    # The specified template must exist
+        await _ensure_credentials_secret_exists(spec, namespace)
+
     next_template = await fetch_model_instance(api.ClusterTemplate, spec.template_name)
     if not next_template:
-        raise kopf.AdmissionError("specified cluster template does not exist", code = 400)
-    # Load the current state of the cluster
-    current_cluster = await fetch_model_instance(api.Cluster, name, namespace = namespace)
-    # If the template is not changing, we are done
-    if current_cluster and current_cluster.spec.template_name == next_template.metadata.name:
+        raise kopf.AdmissionError("specified cluster template does not exist", code=400)
+
+    current_cluster = await fetch_model_instance(api.Cluster, name, namespace=namespace)
+    if current_cluster and (
+        current_cluster.spec.template_name == next_template.metadata.name
+    ):
         return
-    # If we get to here, the template must not be deprecated
+
     if next_template.spec.deprecated:
-        raise kopf.AdmissionError("specified cluster template is deprecated", code = 400)
-    # The rest of the validation only applies to upgrades
+        raise kopf.AdmissionError("specified cluster template is deprecated", code=400)
+
     if not current_cluster:
         return
-    # Load the current template
+
+    await _validate_upgrade(current_cluster, next_template)
+
+    return next_template
+
+
+def _validate_spec_model(spec):
```

Comment:
This is hard to review, its annoying, but lets try split out this change if we can.
I don't mind having a high complexity for now.